Panic in k8sWatcherConverter

Seen in felix's log during a k8sfv test run with `JUST_A_MINUTE=true USE_TYPHA=false make k8sfv-test`

## Expected Behavior
No panics!

## Current Behavior
See this panic about once every 60s in the logs:
```
panic: interface conversion: interface is nil, not resources.Resource

goroutine 102 [running]:
github.com/projectcalico/felix/vendor/github.com/projectcalico/libcalico-go/lib/backend/k8s/resources.(*k8sWatcherConverter).convertEvent(0xc420114e80, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/projectcalico/felix/vendor/github.com/projectcalico/libcalico-go/lib/backend/k8s/resources/watcher.go:118 +0x98
github.com/projectcalico/felix/vendor/github.com/projectcalico/libcalico-go/lib/backend/k8s/resources.(*k8sWatcherConverter).processK8sEvents(0xc420114e80)
	/go/src/github.com/projectcalico/felix/vendor/github.com/projectcalico/libcalico-go/lib/backend/k8s/resources/watcher.go:87 +0x228
created by github.com/projectcalico/felix/vendor/github.com/projectcalico/libcalico-go/lib/backend/k8s/resources.newK8sWatcherConverter
	/go/src/github.com/projectcalico/felix/vendor/github.com/projectcalico/libcalico-go/lib/backend/k8s/resources/watcher.go:43 +0x13d
```
## Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->
* libcalico-go version: 80853cf81b74dbde7b6c55dae0d52cfe8e0001c9
* felix version: 0aba6bc56f0a8990795e2ae491a13bd22b0c4683
* Operating System and version: Ubuntu 16.04
* Link to your project (optional):

--- a/lib/backend/k8s/resources/customresource.go
+++ b/lib/backend/k8s/resources/customresource.go
@@ -312,7 +312,7 @@ func (c *customK8sResourceClient) Watch(ctx context.Context, list model.ListInte
 		r.GetObjectKind().SetGroupVersionKind(c.k8sResourceTypeMeta.GetObjectKind().GroupVersionKind())
 		return c.convertResourceToKVPair(r)
 	}
-	return newK8sWatcherConverter(ctx, toKVPair, k8sWatch), nil
+	return newK8sWatcherConverter(ctx, resl.Kind+" (custom)", toKVPair, k8sWatch), nil
 }
 
 // EnsureInitialized is a no-op since the CRD should be

--- a/lib/backend/k8s/resources/networkpolicy.go
+++ b/lib/backend/k8s/resources/networkpolicy.go
@@ -63,9 +63,10 @@ func NewNetworkPolicyClient(c *kubernetes.Clientset, r *rest.RESTClient) K8sReso
 
 // Implements the api.Client interface for NetworkPolicys.
 type networkPolicyClient struct {
-	clientSet *kubernetes.Clientset
-	crdClient *customK8sResourceClient
-	converter conversion.Converter
+	resourceName string
+	clientSet    *kubernetes.Clientset
+	crdClient    *customK8sResourceClient
+	converter    conversion.Converter
 }
 
 func (c *networkPolicyClient) Create(ctx context.Context, kvp *model.KVPair) (*model.KVPair, error) {
@@ -256,7 +257,7 @@ func (c *networkPolicyClient) Watch(ctx context.Context, list model.ListInterfac
 		}
 		return c.converter.K8sNetworkPolicyToCalico(np)
 	}
-	k8sWatch := newK8sWatcherConverter(ctx, converter, k8sRawWatch)
+	k8sWatch := newK8sWatcherConverter(ctx, "NetworkPolicy (namespaced)", converter, k8sRawWatch)
 
 	calicoWatch, err := c.crdClient.Watch(ctx, list, crdRev)
 	if err != nil {

--- a/lib/backend/k8s/resources/node.go
+++ b/lib/backend/k8s/resources/node.go
@@ -174,7 +174,7 @@ func (c *nodeClient) Watch(ctx context.Context, list model.ListInterface, revisi
 		}
 		return K8sNodeToCalico(k8sNode)
 	}
-	return newK8sWatcherConverter(ctx, converter, k8sWatch), nil
+	return newK8sWatcherConverter(ctx, "Node", converter, k8sWatch), nil
 }
 
 // K8sNodeToCalico converts a Kubernetes format node, with Calico annotations, to a Calico Node.

--- a/lib/backend/k8s/resources/profile.go
+++ b/lib/backend/k8s/resources/profile.go
@@ -150,5 +150,5 @@ func (c *profileClient) Watch(ctx context.Context, list model.ListInterface, rev
 		}
 		return c.converter.NamespaceToProfile(k8sNamespace)
 	}
-	return newK8sWatcherConverter(ctx, converter, k8sWatch), nil
+	return newK8sWatcherConverter(ctx, "Profile", converter, k8sWatch), nil
 }

--- a/lib/backend/k8s/resources/watcher.go
+++ b/lib/backend/k8s/resources/watcher.go
@@ -31,9 +31,15 @@ const (
 	resultsBufSize = 100
 )
 
-func newK8sWatcherConverter(ctx context.Context, converter ConvertK8sResourceToKVPair, k8sWatch kwatch.Interface) api.WatchInterface {
+func newK8sWatcherConverter(
+	ctx context.Context,
+	name string,
+	converter ConvertK8sResourceToKVPair,
+	k8sWatch kwatch.Interface,
+) api.WatchInterface {
 	ctx, cancel := context.WithCancel(ctx)
 	wc := &k8sWatcherConverter{
+		logCxt:     log.WithField("resource", name),
 		converter:  converter,
 		k8sWatch:   k8sWatch,
 		context:    ctx,
@@ -45,6 +51,7 @@ func newK8sWatcherConverter(ctx context.Context, converter ConvertK8sResourceToK
 }
 
 type k8sWatcherConverter struct {
+	logCxt     *log.Entry
 	converter  ConvertK8sResourceToKVPair
 	k8sWatch   kwatch.Interface
 	context    context.Context
@@ -73,9 +80,9 @@ func (crw *k8sWatcherConverter) HasTerminated() bool {
 // Loop to process the events stream from the underlying k8s Watcher and convert them to
 // backend KVPs.
 func (crw *k8sWatcherConverter) processK8sEvents() {
-	log.Info("Watcher process started")
+	crw.logCxt.Info("Watcher process started")
 	defer func() {
-		log.Info("Watcher process terminated")
+		crw.logCxt.Info("Watcher process terminated")
 		crw.Stop()
 		close(crw.resultChan)
 		atomic.AddUint32(&crw.terminated, 1)
@@ -87,23 +94,23 @@ func (crw *k8sWatcherConverter) processK8sEvents() {
 			e := crw.convertEvent(event)
 			select {
 			case crw.resultChan <- e:
-				log.Debug("Kubernetes event converted and sent to backend watcher")
+				crw.logCxt.Debug("Kubernetes event converted and sent to backend watcher")
 
 				// If this is an error event, check to see if it's a terminating one (the
 				// convertEvent method will decide that).  If so, terminate this watcher.
 				if e.Type == api.WatchError {
-					log.WithError(e.Error).Debug("Kubernetes event converted to backend watcher error event")
+					crw.logCxt.WithError(e.Error).Debug("Kubernetes event converted to backend watcher error event")
 					if _, ok := e.Error.(cerrors.ErrorWatchTerminated); ok {
-						log.Info("Watch terminated event")
+						crw.logCxt.Info("Watch terminated event")
 						return
 					}
 				}
 			case <-crw.context.Done():
-				log.Info("Process watcher done event during watch event in kdd client")
+				crw.logCxt.Info("Process watcher done event during watch event in kdd client")
 				return
 			}
 		case <-crw.context.Done(): // user cancel
-			log.Info("Process watcher done event in kdd client")
+			crw.logCxt.Info("Process watcher done event in kdd client")
 			return
 		}
 	}
@@ -114,11 +121,11 @@ func (crw *k8sWatcherConverter) processK8sEvents() {
 func (crw *k8sWatcherConverter) convertEvent(kevent kwatch.Event) api.WatchEvent {
 	var kvp *model.KVPair
 	var err error
-	if kevent.Type != kwatch.Error {
+	if kevent.Type != kwatch.Error && kevent.Type != "" {
 		k8sRes := kevent.Object.(Resource)
 		kvp, err = crw.converter(k8sRes)
 		if err != nil {
-			log.WithError(err).Warning("Error converting Kubernetes resource to Calico resource")
+			crw.logCxt.WithError(err).Warning("Error converting Kubernetes resource to Calico resource")
 			return api.WatchEvent{
 				Type:  api.WatchError,
 				Error: err,
@@ -127,7 +134,7 @@ func (crw *k8sWatcherConverter) convertEvent(kevent kwatch.Event) api.WatchEvent
 	}
 
 	switch kevent.Type {
-	case kwatch.Error:
+	case kwatch.Error, "":
 		// An error directly from the k8s watcher is a terminating event.
 		return api.WatchEvent{
 			Type: api.WatchError,

--- a/lib/backend/k8s/resources/workloadendpoint.go
+++ b/lib/backend/k8s/resources/workloadendpoint.go
@@ -192,5 +192,5 @@ func (c *WorkloadEndpointClient) Watch(ctx context.Context, list model.ListInter
 		}
 		return c.converter.PodToWorkloadEndpoint(k8sPod)
 	}
-	return newK8sWatcherConverter(ctx, converter, k8sWatch), nil
+	return newK8sWatcherConverter(ctx, "Pod", converter, k8sWatch), nil
 }

--- a/lib/backend/syncersv1/updateprocessors/networkpolicyprocessor_test.go
+++ b/lib/backend/syncersv1/updateprocessors/networkpolicyprocessor_test.go
@@ -233,7 +233,6 @@ var _ = Describe("Test the NetworkPolicy update processor", func() {
 		Expect(rulev1.NotICMPCode).To(Equal(&encode))
 		Expect(rulev1.NotICMPType).To(Equal(&entype))
 
-
 		Expect(rulev1.SrcNets).To(Equal([]*cnet.IPNet{mustParseCIDR("10.100.1.1/32")}))
 		Expect(rulev1.SrcSelector).To(Equal("kns.namespacelabel1 == 'value1'"))
 		Expect(rulev1.SrcPorts).To(Equal([]numorstring.Port{port443}))

